src/client_side_request.cc

```diff
@@ -1794,9 +1794,13 @@ ClientHttpRequest::doCallouts()
             ACLFilledChecklist ch(NULL, request, NULL);
             ch.src_addr = request->client_addr;
             ch.my_addr = request->my_addr;
-            nfmark_t mark = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
-            if (mark)
+            if (const auto mc = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch)) {
+                // Clear the bits given by the mask in existing CONNMARK
+                // and OR it with a new mark.
+                nfmark_t mark = getConn()->clientConnection->nfmark & ~(mc->nfmask);
+                mark |= mc->nfmark;
```

Comment:
Again, we muast apply nfmask to nfmark. Use mc->maskedMark() discussed elsewhere.

Reply:
AFAIK, set-mark in iptables works like this: ctmark = (ctmark AND NOT mask) OR value.

Reply:
Documentation says otherwise: `If a mask is specified then only those bits set in the mask are modified`. However, the documentation may lie. If you can test what iptables do, please test.

Reply:
According to [this]https://github.com/theojulienne/iptables/blob/master/extensions/libxt_MARK.c#L126 and [this]https://github.com/torvalds/linux/blob/9cb9c07d6b0c5fd97d83b8ab14d7e308ba4b612f/net/netfilter/xt_connmark.c#L92, the mask is applied to original mark. Also, if I understand correctly, netfilter expresses set-mark through XOR.

Reply:
Actually, iptables has very odd documentation :) The statement can be also treated as 'if we have a mask, we zero out corresponding bits in a source mark and then apply a new mark to it'.

Reply:
We've done some tests when clientside_mark ACL was added and everything worked as expected.

Reply:
However, we can run additional tests on monday, if needed. Any more remarks on this PR?

Reply:
Sorry for not being clear: The tests I mentioned were concerning iptables, not Squid. I understand that your Squid implementation matches your expectations, and that is great, but that is not the issue here. The iptables questions are:

1. Does `--mark 0xBF/0xF` match an original/preset 0xAF mark?
2. Given no/zero original/preset mark, does `--set-mark 0xFF/0xF` set a 0xF mark or a 0xFF mark?

I do not know the exact test commands for figuring out the above, but I assume that you do.

> Docs: If a mask is specified then only those bits set in the mask are modified

> The statement can be also treated as 'if we have a mask, we zero out corresponding bits in a source mark and then apply a new mark to it'.

I disagree that it is a reasonable interpretation because the docs talk about restricting _modification _ and modification only makes sense if applied to the connection mark, not any configured value (which is a _constant_ in this context). However, again, iptables documentation may be lying or be misleading -- we should not rely on it (alone).

BTW, I would agree with your interpretation if you replace "it" with "them" or "those bits".

In fact, even relying on the iptables behavior/code _alone_ is dangerous -- it could be buggy when it comes to unusual configurations such as 0xBF/0xF. There should be some common, well-known approach to these issues that all tools, including iptables and Squid should follow instead. Unfortunately, I do not know about that approach because I am not a packet filtering expert. I assume that you do not know it either (because you have not referred to it yet). And nobody else has chimed in. Thus, we are trying to make progress by testing iptables and looking at iptables code and documentation. This is not the best or an efficient way to do it, but it may still result in a well-understood (if not necessarily correct) commit.

> Any more remarks on this PR?

I do not know yet. If you are done with addressing the previous comments, then I will review the changes once again after this 0xBF/0xF issue is fully addressed.

Reply:
A quick test shows:
--mark 0xBF/0xF does not match 0xAF
--set-mark 0xFF/0xF sets 0xFF

The test rules were:
```
iptables -t nat -A OUTPUT -p tcp --dport 80 -j MARK --set-mark 0xAFk
iptables -A OUTPUT -m mark 0xBF/0xF -j DROP
```

Reply:
I've addressed all previous comments in current version (except of a marking logic that is currently beeing discussed).

Reply:
@hno, I noticed you authored MARA's connection marking Linux [code](https://github.com/torvalds/linux/blob/9cb9c07d6b0c5fd97d83b8ab14d7e308ba4b612f/net/netfilter/xt_connmark.c#L30). This PR adds `/mask` support to Squid's existing [clientside_mark](http://www.squid-cache.org/Doc/config/clientside_mark/) directive. Assuming that we are implementing an iptables-like `--set-mark` semantics (and _not_ `--set-xmark` semantics!), which of the following options for treating "excessive" mark values (as in 0xAB/0xF) would you recommend?

1. Excessive bits should be ignored: Setting 0xAB/0xF means setting 0xB/0xF.
2. Excessive bits should be honored: Setting 0xAB/0xF means setting 0xAB/0xFF.
3. Excessive bits is a configuration error. Different reasonable people may expect different outcomes when setting 0xAB/0xF.

Same questions apply to matching 0xAB/0xF when using a clientside_mark ACL, but I assume the answers to the "matching" questions would be the same as the answers to the "setting" questions above.

Reply:
@gozzy, thank you for testing those cases! I believe both tests imply that 0xAB/0xF means 0xAB/0xFF to iptables (which is what your latest code implements). Let's continue to use that semantics at least until an expert chimes in. Please ignore my maskedValue() suggestions. 

I will check the latest code shortly.